packages/widgets/src/blocks/legacy-widget/edit/utils.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useLayoutEffect, useState } from '@wordpress/element';
+import { useResizeObserver } from '@wordpress/compose';
+
+/**
+ * Positions an element by the top of an objective element if it can fit or
+ * otherwise as close as possible within the window.
+ *
+ * @param {Element} objective    Element to take position from.
+ * @param {Object}  $1           Options.
+ * @param {boolean} $1.isEnabled Whether to apply the hook or not.
+ *
+ * @return {Object} A ref and resize observer element.
+ */
+export function useBudgeTopBy( objective, { isEnabled } ) {
+	const subjectRef = useRef();
+	const objectiveRef = useRef();
+	const [
+		{ start: boundsStart, end: boundsEnd } = {},
+		setBounds,
+	] = useState();
+	const boundsUpdaterRef = useRef();
+
+	const hasSubject = !! subjectRef.current;
+	const hasObjective = !! objective;
+	const objectiveHasChanged = objectiveRef.current !== objective;
+	objectiveRef.current = objective;
+
+	// Defines the bounds updating function a single time
+	if ( ! boundsUpdaterRef.current ) {
+		boundsUpdaterRef.current = () => {
+			const { defaultView } = subjectRef.current.ownerDocument;
+			setBounds( { start: 0, end: defaultView.innerHeight } );
+		};
+	}
+
+	// Sets the bounds height and handles window resizes to update it
+	useLayoutEffect( () => {
+		if ( ! hasSubject || ! hasObjective || ! isEnabled ) {
+			return;
+		}
+		boundsUpdaterRef.current();
+		const { defaultView } = objective.ownerDocument;
+		const updateBounds = boundsUpdaterRef.current;
+		defaultView.addEventListener( 'resize', updateBounds );
+		return () => defaultView.removeEventListener( 'resize', updateBounds );
+	}, [ hasObjective, hasSubject, isEnabled ] );
```

Comment:
There's potential to break this down into smaller hooks, e.g. `const [ size, [ elementRef ] ] = useViewportSize()`.

Also noticed it subscribes to the objective viewport resize event, but returns the subject size. I think it should always use the objective as that's the element that needs to remain in bounds.

Reply:
I had a look a breaking into smaller hooks but wasn't convinced it was helpful for now (adds more lines). Though, there are around, I think, nine instance elsewhere in the repo a `useViewportSize` hook could be utilized so maybe it's a hook we should have.

I did as suggested for the listener.